src/EventStore.Core/Authorization/PolicyEvaluator.cs

```diff
@@ -1,36 +1,40 @@
 ﻿using System;
+using System.Linq;
 using System.Security.Claims;
 using System.Threading;
 using System.Threading.Tasks;
 using EventStore.Plugins.Authorization;
 
 namespace EventStore.Core.Authorization {
-	public class PolicyEvaluator : IPolicyEvaluator {
+	public class MultiPolicyEvaluator : IPolicyEvaluator {
 		private static readonly AssertionInformation DeniedByDefault =
 			new AssertionInformation("default", "denied by default", Grant.Deny);
 
-		private readonly ReadOnlyPolicy _policy;
+		private readonly IPolicySelector[] _policySelectors;
 		private readonly PolicyInformation _policyInfo;
 
-		public PolicyEvaluator(ReadOnlyPolicy policy) {
-			_policy = policy;
-			_policyInfo = policy.Information;
+		public MultiPolicyEvaluator(IPolicySelector[] policySelectors) {
+			_policySelectors = policySelectors;
+			_policyInfo = new PolicyInformation("multi", 1, DateTimeOffset.MinValue);
 		}
 
 		public ValueTask<EvaluationResult>
 			EvaluateAsync(ClaimsPrincipal cp, Operation operation, CancellationToken ct) {
 			var evaluation = new EvaluationContext(operation, ct);
 
-			if (_policy.TryGetAssertions(operation, out var assertions))
-				while (!assertions.IsEmpty && evaluation.Grant != Grant.Deny) {
-					if (ct.IsCancellationRequested) break;
-					var assertion = assertions.Span[0];
-					assertions = assertions.Slice(1);
-					var evaluate = assertion.Evaluate(cp, operation, _policyInfo, evaluation);
-					if (!evaluate.IsCompleted)
-						return EvaluateAsync(evaluate, cp, operation, assertions, evaluation, ct);
-				}
-
+			foreach (var policySelector in _policySelectors) {
+				var policy = policySelector.Select();
+				var policyInfo = policy.Information;
+				if (policy.TryGetAssertions(operation, out var assertions))
+					while (!assertions.IsEmpty && evaluation.Grant != Grant.Deny) {
+						if (ct.IsCancellationRequested) break;
+						var assertion = assertions.Span[0];
+						assertions = assertions.Slice(1);
+						var evaluate = assertion.Evaluate(cp, operation, policyInfo, evaluation);
+						if (!evaluate.IsCompleted)
+							return EvaluateAsync(evaluate, cp, operation, assertions, evaluation, ct);
+					}
```

Comment:
it appears that there could be two bugs here:
1. when `assertion.Evaluate()` is processed synchronously: if `evaluation.Grant == Grant.Deny` , we don't break out of the `foreach` loop and return the result, but instead, we continue to process the next policy selector

2. when `assertion.Evaluate()` is processed asynchronously: we don't process the next policy selectors at all (e.g if the result is `Grant.Unknown` in `EvaluateAsync`, it'll be translated into a `DeniedByDefault` error but we should instead be processing the next policy selectors)

Reply:
it would be much simpler to use `async`/`await` here, although i'm not sure if the performance difference would be significant (given that it's on the hot path). we could measure the difference to be sure it doesn't have any negative impact

Reply:
Thanks, I've updated `MultiPolicyEvaluator` to use `async`/`await`, and added tests for it

I did some benchmarks with the MicroBenchmarks project, and it looks like there's no significant performance difference between using `async`/`await` and not. So it makes sense to keep the code simple

When the call is made synchronously:

| Method             | Mean     | Error   | StdDev  | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------- |---------:|--------:|--------:|------:|--------:|-------:|----------:|------------:|
| Optimised_NoAwait  | 138.7 ns | 2.69 ns | 2.64 ns |  1.00 |    0.00 | 0.0739 |     464 B |        1.00 |
| NonOptimised_Await | 106.1 ns | 2.08 ns | 2.92 ns |  0.76 |    0.03 | 0.0612 |     384 B |        0.83 |

When the caller is using `async`/`await` themselves

| Method             | Mean     | Error   | StdDev  | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------- |---------:|--------:|--------:|------:|--------:|-------:|----------:|------------:|
| Optimised_NoAwait  | 157.5 ns | 3.15 ns | 6.86 ns |  1.00 |    0.00 | 0.0739 |     464 B |        1.00 |
| NonOptimised_Await | 135.7 ns | 2.38 ns | 2.11 ns |  0.86 |    0.03 | 0.0610 |     384 B |        0.83 |

Reply:
great, thanks!